### PR TITLE
[MM-18768] Remove deprecated lifecycle methods from channel info

### PR DESCRIPTION
**Pull request by goku321**
#### Summary
Remove deprecated lifecycle methods from channel_info and edit_channel
#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/12351

### CODE REVIEW

File: app/screens/channel_info/channel_info.js [hunk 1]
```diff
@@ -82,6 +82,7 @@ export default class ChannelInfo extends PureComponent {
             isFavorite: props.isFavorite,
             isMuted: props.isChannelMuted,
             ignoreChannelMentions: props.ignoreChannelMentions,
+            prevPropsUserId: props.currentUserId,
```
Comment (enahum) [inline app/screens/channel_info/channel_info.js]: why do you need the `currentUserId` in the state?

File: app/screens/channel_info/channel_info.js [hunk 2]
```diff
@@ -90,27 +91,37 @@ export default class ChannelInfo extends PureComponent {
...
         }
 
-        let ignoreChannelMentions = this.state.ignoreChannelMentions;
+        let ignoreChannelMentions = state.ignoreChannelMentions;
         if (ignoreChannelMentions !== nextProps.ignoreChannelMentions) {
             ignoreChannelMentions = nextProps.ignoreChannelMentions;
         }
 
-        this.setState({isFavorite, isMuted, ignoreChannelMentions});
+        if (nextProps.currentUserId !== state.prevPropsUserId) {
```
Comment (enahum) [inline app/screens/channel_info/channel_info.js]: not following, why returning the state only when the currentUserId changes? this won't even happen as the currentUserId does not changes while you are logged in with the same user.

File: app/screens/edit_channel/edit_channel.js [hunk 3]
```diff
@@ -92,6 +92,7 @@ export default class EditChannel extends PureComponent {
         this.state = {
             error: null,
             updating: false,
+            updateChannelRequest: this.props.updateChannelRequest,
```
Comment (enahum) [inline app/screens/edit_channel/edit_channel.js]: we should remove this property as it is being deprecated, but I believe that is out of scope of this PR.

File: app/screens/channel_info/channel_info.js (hunk 1 above)
Comment (goku321) [reply app/screens/channel_info/channel_info.js]: Reason for keeping this here was to update `state` from `props` only when the `currentUserId` changes and in other cases `state` update would be handled by the `handler`. But, as you mentioned in the other comment, I think this approach is not correct

File: app/screens/channel_info/channel_info.js (hunk 2 above)
Comment (goku321) [reply app/screens/channel_info/channel_info.js]: Should I just simply return the state here? But then, state updates would be lost (I think so).

Comment (hmhealey) [reply app/screens/channel_info/channel_info.js]: Perhaps we could just remove `getDerivedStateFromProps` entirely? We already set the state from the props when this screen is opened, and the click handlers for each of those settings update the state when the user taps on them, so the only case we'd really lose out on is if the user updates one of those from another device while this screen is open which is virtually never going to happen

File: app/screens/edit_channel/edit_channel.js [hunk 4]
```diff
@@ -114,32 +115,42 @@ export default class EditChannel extends PureComponent {
...
-        if (this.props.updateChannelRequest !== updateChannelRequest) {
+        if (state.updateChannelRequest !== updateChannelRequest) {
+            const newState = {
+                error: null,
+                updating: true,
+            };
+
             switch (updateChannelRequest.status) {
             case RequestStatus.STARTED:
                 this.emitUpdating(true);
```
Comment (hmhealey) [inline app/screens/edit_channel/edit_channel.js]: `this` is no longer defined, so this will need to be moved to the `componentDidUpdate`. I think the same thing needs to happen for the calls to `EventEmitter` and such below as well

File: app/screens/edit_channel/edit_channel.js (hunk 3 above)
Comment (hmhealey) [reply app/screens/edit_channel/edit_channel.js]: Yeah, it's out of scope. I believe we're only removing the request status that isn't being used anywhere, so no changes are needed here.

File: app/screens/edit_channel/edit_channel.js (hunk 4 above)
Comment (goku321) [reply app/screens/edit_channel/edit_channel.js]: Hi @hmhealey thanks for your suggestions. I have a question: do we need `getDerivedStateFromProps` here because there can be a case when a user is trying to edit a channel and someone has already updated the channel?

Comment (hmhealey) [reply app/screens/edit_channel/edit_channel.js]: No, I don't think we need to handle that case here. `updateChannelRequest` only tracks requests made from this device, so we don't see anything from other users even if we use `getDerivedStateFromProps`

File: app/screens/edit_channel/edit_channel.js [hunk 5]
```diff
@@ -114,30 +115,49 @@ export default class EditChannel extends PureComponent {
...
-        if (this.props.theme !== nextProps.theme) {
-            setNavigatorStyles(this.props.componentId, nextProps.theme);
-        }
-
+    static getDerivedStateFromProps(nextProps, state) {
         const {updateChannelRequest} = nextProps;
 
-        if (this.props.updateChannelRequest !== updateChannelRequest) {
+        if (state.updateChannelRequest !== updateChannelRequest) {
+            const newState = {
```
Comment (hmhealey) [inline app/screens/edit_channel/edit_channel.js]: You need to update `this.state.updateChannelRequest` here as well. Otherwise, this looks good

Comment (goku321) [reply app/screens/edit_channel/edit_channel.js]: My bad, will fix this.

**Comment by goku321** [comment]
Thanks @hmhealey and @enahum for your suggestions and patience

File: app/screens/channel_info/channel_info.js [hunk 6]
```diff
@@ -90,27 +90,10 @@ export default class ChannelInfo extends PureComponent {
...
 
-    componentWillReceiveProps(nextProps) {
-        if (this.props.theme !== nextProps.theme) {
-            setNavigatorStyles(this.props.componentId, nextProps.theme);
+    componentDidUpdate(prevProps) {
+        if (prevProps.theme !== this.props.theme) {
+            setNavigatorStyles(prevProps.componentId, this.props.theme);
         }
-
-        let isFavorite = this.state.isFavorite;
```
[the remark this answers is not in the document]
Comment (goku321) [reply app/screens/channel_info/channel_info.js]: Sure 👍

**Comment by prapti** [comment]
@hmhealey @enahum Can either of you help with a build to test this PR? Thank you!

**Comment by enahum** [comment]
@prapti to be honest I don’t this needs a build, perhaps running it on a Sim is enough?

**Comment by prapti** [comment]
> @prapti to be honest I don’t this needs a build, perhaps running it on a Sim is enough?

@enahum I don't have sims set up on my end. In that case, would it be enough if only devs approved?

**Comment by enahum** [comment]
@prapti I'll build the PR. be on the lookout for the builds in the QA channel

**Comment by prapti** [comment]
Sounds good. Thank you @enahum !

**Comment by prapti** [comment]
@hmhealey this one is ready to be merged!
